### PR TITLE
Modify interface to support HttpClientFactory users

### PR DESCRIPTION
**Pull request by thecontrarycat**
PR for #491

### CODE REVIEW

File: InterfaceStubGenerator.Core/GeneratedInterfaceStubTemplate.mustache [hunk 1]
```diff
@@ -63,7 +62,7 @@ namespace {{Namespace}}
         {
 {{#IsRefitMethod}}    
             var arguments = new object[] { {{ArgumentList}} };
-            var func = methodImpls.GetOrAdd({{#MethodTypeParameterNames}}${{/MethodTypeParameterNames}}"{{Name}}{{#MethodTypeParameterNames}}<{{.}}>{{/MethodTypeParameterNames}}({{ArgumentListWithTypes}})", _ => requestBuilder.BuildRestResultFuncForMethod("{{Name}}", new Type[] { {{ArgumentTypesList}} }{{#MethodTypeParameterList}}, new Type[] { {{.}} }{{/MethodTypeParameterList}}));
+            var func = requestBuilder.BuildRestResultFuncForMethod("{{Name}}", new Type[] { {{ArgumentTypesList}} }{{#MethodTypeParameterList}}, new Type[] { {{.}} }{{/MethodTypeParameterList}});
```
Comment (onovotny) [inline InterfaceStubGenerator.Core/GeneratedInterfaceStubTemplate.mustache]: We can't do this....this means we're running potentially expensive reflection code on every method invocation and not caching the result.

I'd need to see a benchmark comparing the two before removing this cache.

**Comment by onovotny** [comment]
I may have spoken a little too soon in my code review comment as I was going top-down and didn't see the new cache.

Given the structural changes, can you please describe what you changed and how it's working & caching things?

File: Refit/CachedRequestBuilderImplementation.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        string GetArgumentString(Type[] parameterTypes)
+        {
+            if (parameterTypes == null || parameterTypes.Length == 0)
+            {
+                return "";
+            }
+
+            return parameterTypes.Select(t => t.Name).Aggregate((s1, s2) => s1 + ", " + s2);
```
Comment (bennor) [inline Refit/CachedRequestBuilderImplementation.cs]: `string.Join(", ", parameterTypes.Select(t => t.Name)` would be a little clearer here. There's another one like this in the next method too.

File: Refit/RequestBuilderImplementation.cs [hunk 3]
```diff
@@ -27,7 +27,7 @@ partial class RequestBuilderImplementation : IRequestBuilder
         readonly ConcurrentDictionary<CloseGenericMethodKey, RestMethodInfo> interfaceGenericHttpMethods;
         readonly JsonSerializer serializer;
         readonly RefitSettings settings;
-        readonly Type targetType;
+        public Type targetType { get; }
```
Comment (onovotny) [inline Refit/RequestBuilderImplementation.cs]: Since this is a property, change to PascalCase

Comment (thecontrarycat) [reply Refit/RequestBuilderImplementation.cs]: Good spot.  Fixing now

**Comment by onovotny** [comment]
Haven't forgotten about this. Will try to get this reviewed and merged in the next few days.

**Comment by onovotny** [comment]
@thecontrarycat This looks good, thanks! One thing that would be helpful though is a doc update that shows how to use this with `IHttpClientFactory`. Can you please add that and I'll complete the merge?

**Comment by thecontrarycat** [comment]
@onovotny I've added a note to the documentation showing how you can use it with HttpClientFactory

**Comment by onovotny** [comment]
@thecontrarycat Thanks, that makes sense.

In light of the pattern in the docs, what would we need to expose a single extension method (or two with overloads) to add Refit? Something like `.AddRefit<T>(...)` with the settings, or whatever?

**Comment by thecontrarycat** [comment]
I think it would be complicated to do a single `IServiceCollection.AddRefit<T>()` method because we'd need to implement the whole `HttpClient` configuration part (there are additional lines you can put between the `AddHttpClient` and `AddTypedClient` methods to customise the handlers used by the client).

I did start thinking that we could make `IRequestHandler` become `IRequestHandler<T>` and we could then replace the `AddTypedClient` call with an `AddRefitClient<T>` call where we could singleton register the appropriate `IRequestHandler<T>` in the same call.  

Refactoring the IRequestHandler was a larger job than I wanted to do unless you think it makes sense.

**Comment by thecontrarycat** [comment]
I did that refactor.  `IRequestBuilder` and `IRequestBuilderInternal` have become `IRequestBuilder<T>` and `IRequestBuilder`, with the generic version being used during creation.  This allows us to have a single extension method to configure Refit in the `HttpClientFactory` world (see updated docs).

File: Refit/HttpClientFactoryExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿#if NETSTANDARD2_0
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Refit
+{
+    public static class HttpClientFactoryExtensions
+    {
+        public static IHttpClientBuilder AddRefitClient<T>(this IServiceCollection services) where T : class
```
Comment (onovotny) [inline Refit/HttpClientFactoryExtensions.cs]: I think we need an overload that can specify the refit settings. Can either be as another parameter or as an `Action<RefitSettings>` where the method news it up and passes it to the user for setting?

**Comment by onovotny** [comment]
This looks really great, I love having an easy one-liner to add in the startup that "does the right thing"

**Comment by thecontrarycat** [comment]
Added the overload for `RefitSettings`.  It's not an `Action<RefitSettings>` because the `IRequestBuilder<T>` will be a singleton and therefore there's no ability to change the value at runtime.

**Comment by onovotny** [comment]
Thank for this!

**Comment by onovotny** [comment]
@thecontrarycat I know I already merged it, but any chance you could add a test or two that covers what the extension method is supposed to do and any tests that cover the new behavior?

**Pull request merged**